### ISSUE TITLE
Treat dash-connected words as a single word for citekey generation

### ISSUE THREAD

**Issue by ookimi** [retorquere/zotero-better-bibtex#619]
I am trying to run a postfix script with simple search-and-replace on the generated bibtex citation key - but for the love of God I can't find how to access it.
I would have guessed "this.item.citekey" - but this pattern doesn't work:

this.item.citekey.replace('a', 'b');

**Comment by retorquere** [comment]
It's available as `this.item.__citekey__` (I've added the underscores to prevent potential conflicts with existing or upcoming Zotero reference item fields), and you'd have to do something like

```
this.item.__citekey__ = this.item.__citekey__.replace('a', 'b')
```

because replace doesn't change its target, but please keep in mind that this will not change the cite key as it is used anywhere else, it's *only* during final export. In the UI you will see the unmodified citekey.

Also keep in mind that the reference so produced is cached, so the postscript is ran only once regardless of how many exports you do until you change the reference in Zotero. Something like

```
this.item.__citekey__ = this.item.__citekey__ + new Date()
```

(if you were to want something so odd) will not change on every export.

This is all untested and I'm a little nervous about the citekey being changed in the postscript TBH. I've never tested how modifying the citekey will affect the caching system; I *think* it should be alright but that's about the extent of my certitude.

What do you want to correct in the citekey?

**Comment by ookimi** [comment]
Hmmm. Not sure how to proceed then...

I am trying to get rid of all punctuation in titles, but leave "-". My papers have a lot of long hyphenated words like "causative-inchoative" and "situation-bound" - and "situationbound" is rather hard to read. Since nopunc in the citekey generator removes it all, I thought I would try to remove all other punctuation marks except "-" in the postscript.

Or is there a magic setting that defines what gets removed by nopunc?

**Comment by retorquere** [comment]
No magic setting, no... I'd have to think about this one. `nopunct` removes the `punctuation` character class as defined by the unicode spec, which includes `punctuation-dash`. 

Having 

```
var re = Zotero.Utilities.XRegExp('\\p{Pc}|\\p{Pe}|\\p{Pf}|\\p{Pi}|\\p{Po}|\\p{Ps}', 'g')
this.item.__citekey__  = Zotero.Utilities.XRegExp.replace(this.item.__citekey__, re, '', 'all')
```

might do what you want... but again, I'd have to dive into the caching system to make sure that doesn't give wacky results elsewhere. I think not, as what I'm seeing so far caches the citekey but doesn't use it as a lookup key, and as the citekey is not guaranteed unique (legitimate duplicates may exist across libraries, and if you pin citekeys I don't prevent you from pinning duplicates) it really shouldn't, but I'm not 100% sure. Except of course you won't see it in the UI, and the Quick Copy settings don't run the postscript, which affects key drag-and-drop.

I'm trying to think of a way that works everywhere but doesn't lead to a (further) proliferation of preferences.

**Comment by ookimi** [comment]
in settings we trust :) 

Too bad there's no prettier way to do this. I'd say this is a linguistic bug on the side of JabRef; would be cool if you could fix it somehow.

The thing is, nopunc is basically correct - except that it ignores this one case when punctuation ("-") is a part of word ("winnie-the-pooh"), and not something extra and optional. "-" should be stripped when it is a long dash "to be or not to be - this is the question", but kept when it's snug in the middle of a word. Too bad JabRef people didn't think about this :(

**Comment by retorquere** [comment]
That sounds reasonable, actually. AAMOF I think I already treat words with dashes as single words. I'll look into it.

**Comment by retorquere** [comment]
So, to get back to this; you'd want winnie-the-pooh not only treated as a single word, but to keep the dashes in the citekey too?

**Comment by retorquere** [comment]
Also, can you get me a test case by right-clicking a reference where you'd want this and submitting a BBT error report?

**Comment by ookimi** [comment]
Yeah, I'd say that's the most sensible thing to do.
Smith_et_al_Winnie-the-pooh_as_the_center_of_narrative_gravity_in_a_nonsensical_paper_title :)

**Comment by ookimi** [comment]
Ok, just submitted a report (Q4BCDGJJ)
The paper is titled "A Frame Tracking Model for Memory-Enhanced Dialogue Systems"

**Comment by retorquere** [comment]
That currently gets me

```
@article{SchulzZumerEtAl_2017_Frame_Tracking_Model_for_MemoryEnhanced_Dialogue_Systems,
  author = {Schulz, Hannes and Zumer, Jeremie and Asri, Layla El and Sharma, Shikhar},
  date = {2017},
  journaltitle = {arXiv preprint arXiv:1706.01690},
  keywords = {_tablet_modified},
  title = {A {{Frame Tracking Model}} for {{Memory}}-{{Enhanced Dialogue Systems}}},
  url = {https://arxiv.org/abs/1706.01690},
  urldate = {2017-08-17}
}
```

what would you want it to be instead? `SchulzZumerEtAl_2017_Frame_Tracking_Model_for_Memory-Enhanced_Dialogue_Systems`?

**Comment by ookimi** [comment]
yep, SchulzZumerEtAl_2017_Frame_Tracking_Model_for_Memory-Enhanced_Dialogue_Systems.
it's readable without the dash in this example, but not in many others.

**Comment by retorquere** [comment]
BTW, in that sample you have the key pinned, and you have a pattern that uses `skipwords` but you've cleared the skipwords list in your preferences.

**Comment by retorquere** [comment]
So the question is now -- which of the [Pc](http://www.fileformat.info/info/unicode/category/Pc/list.htm)/[Pd](http://www.fileformat.info/info/unicode/category/Pd/list.htm) set to keep/skip/replace with an ASCII-dash (`-`).

**Comment by ookimi** [comment]
Oh god... how about just replacing all of them with ASCII-dash? double and triple dashes can just turn to -- and ---.

And yeah, I know about the skipwords, I just forgot to remove it from the pattern.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 4737](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.36.4737.issue-619.xpi) ("fixes #619").

**Comment by retorquere** [comment]
Replacing them all with ASCII-dash is tons simpler, so I've gone with that. Try 4737.

**Comment by ookimi** [comment]
Just tested 4737 - the hyphens are intact now. Thanks!

**Comment by retorquere** [comment]
Alright, will be merged into the next release, somewhere in the next few days.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 4743](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.37.4743.issue-619.xpi) ("Merge branch 'master' into 619").
